server/pypi/packages/scipy/patches/chaquopy.patch

```diff
@@ -100,21 +119,42 @@ diff -aur src-orig/scipy/__init__.py src/scipy/__init__.py
  
  # We first need to detect if we're being called as part of the SciPy
  # setup procedure itself in a reliable manner.
-@@ -121,6 +125,10 @@
-     import sys as _sys
-     _sys.stderr.write('Running from SciPy source directory.\n')
-     del _sys
+@@ -129,6 +130,10 @@ if __SCIPY_SETUP__:
+     import sys
+     sys.stderr.write('Running from SciPy source directory.\n')
+     del sys
 +
 +    # Chaquopy: the module namespace was originally empty in this mode, but the following
 +    # things are neded by the scikit-learn build.
 +    from scipy.version import version as __version__
  else:
      try:
          from scipy.__config__ import show as show_config
-diff -aur src-orig/scipy/_lib/_testutils.py src/scipy/_lib/_testutils.py
---- src-orig/scipy/_lib/_testutils.py	2023-09-20 11:08:01.940923505 +0200
-+++ src/scipy/_lib/_testutils.py	2023-09-21 13:13:08.763896205 +0200
-@@ -86,9 +86,10 @@
+diff --git a/scipy/_lib/_c99compat.h b/scipy/_lib/_c99compat.h
+index 7d285137d..f0ce4f63f 100644
+--- a/scipy/_lib/_c99compat.h
++++ b/scipy/_lib/_c99compat.h
+@@ -120,9 +120,12 @@
+         /* Hopefully fail in fewer cases */
+ 
+         /* For compilers which aren't MSVC and haven't defined isnan */
+-        #ifndef isnan
+-            #define isnan(x) ((x) != (x))
+-        #endif
++	/* Chaquopy: Prevent C/C++ macro collision */
++	#ifndef __cplusplus
++	    #ifndef isnan
```

Comment:
Why is this patch necessary – what error message do you get without it?

And why is `__cplusplus` the appropriate way of doing it? Is this file included in any C++ code? If not, then the `__cplusplus` check is just equivalent to disabling the code completely.

If the patch is necessary, then the inner #ifndef should be left at its original indentation to make the patch easier to read.

Reply:
My understanding is that this preprocessor guard is necessary since this is a C header file that is being compiled with C++ at some point during the Chaquopy build process. The build would fail without it.